runtime/health.go

```diff
@@ -58,5 +50,183 @@ func (r *Runtime) Health(ctx context.Context) (*Health, error) {
+		canScaleToZero = canScaleToZero || mayBeScaledToZero
+		release()
+
+		// only use cached health if the OLAP can be scaled to zero
+		if cachedHealth != nil && mayBeScaledToZero {
+			mvHealth, ok := cachedHealth.MetricsViews[mv.Meta.Name.Name]
+			if ok && mvHealth.Version == mv.Meta.StateVersion {
+				res.MetricsViews[mv.Meta.Name.Name] = mvHealth
+				continue
+			}
+		}
+		_, err = r.Resolve(ctx, &ResolveOptions{
+			InstanceID:         ctrl.InstanceID,
+			Resolver:           "metricsview_time_range",
+			ResolverProperties: map[string]any{"metrics_view": mv.Meta.Name.Name},
+			Args:               map[string]any{"priority": 10},
+			Claims:             &SecurityClaims{SkipChecks: true},
+		})
+
+		mvHealth := metricsViewHealth{Version: mv.Meta.StateVersion}
+		if err != nil {
+			mvHealth.err = err.Error()
+		}
+		res.MetricsViews[mv.Meta.Name.Name] = mvHealth
+	}
+
+	if !canScaleToZero {
+		return res, nil
+	}
+
+	// save to cache
+	res.ControllerVersion = ctrl.catalog.version
+	bytes, err := json.Marshal(res)
+	if err != nil {
+		return nil, err
+	}
+
+	catalog, release, err := r.Catalog(ctx, instanceID)
+	if err != nil {
+		return nil, err
+	}
+	defer release()
+
+	err = catalog.UpsertInstanceHealth(ctx, &drivers.InstanceHealth{
+		InstanceID: instanceID,
+		Health:     bytes,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (r *Runtime) cachedInstanceHealth(ctx context.Context, instanceID string, ctrlVersion int64) (*InstanceHealth, bool) {
+	catalog, release, err := r.Catalog(ctx, instanceID)
+	if err != nil {
+		return nil, false
+	}
+	defer release()
+
+	cached, err := catalog.FindInstanceHealth(ctx, instanceID)
+	if err != nil {
+		return nil, false
+	}
+
+	c := &InstanceHealth{}
+	err = json.Unmarshal(cached.Health, c)
+	if err != nil || ctrlVersion != c.ControllerVersion {
+		return nil, false
+	}
+	return c, true
+}
+
+func (r *Runtime) pingRepo(ctx context.Context, instanceID string) error {
+	repo, rr, err := r.Repo(ctx, instanceID)
+	if err != nil {
+		return err
+	}
+	defer rr()
+	h, ok := repo.(drivers.Handle)
+	if !ok {
+		return errors.New("unable to ping repo")
+	}
+	return h.Ping(ctx)
+}
+
+func (r *Runtime) pingOLAP(ctx context.Context, olap drivers.OLAPStore) error {
+	h, ok := olap.(drivers.Handle)
+	if !ok {
+		return errors.New("unable to ping olap")
+	}
+	return h.Ping(ctx)
+}
+
+func (h *InstanceHealth) To() *runtimev1.InstanceHealth {
+	if h == nil {
+		return nil
+	}
+	r := &runtimev1.InstanceHealth{
+		ControllerError:     h.Controller,
+		RepoError:           h.Repo,
+		OlapError:           h.OLAP,
+		ParseErrorCount:     int32(h.ParseErrCount),
+		ReconcileErrorCount: int32(h.ReconcileErrCount),
+	}
+	r.MetricsViewErrors = make(map[string]string, len(h.MetricsViews))
+	for k, v := range h.MetricsViews {
+		if v.err != "" {
+			r.MetricsViewErrors[k] = v.err
+		}
+	}
+	return r
+}
+
+type metricsViewHealth struct {
+	err     string
+	Version int64
 }
```

Comment:
1. `InstanceHealth` is an exported type, so any types it depends upon should also be.
2. Move this definition up under the `InstanceHealth` definition.
3. The `runtime` package is a large scope with a lot of functionality, so consider calling the type something like `InstanceHealthMetricsViewError`.